src/mesh/RTree.hpp

```diff
@@ -1,21 +1,161 @@
 #pragma once
 
 #include <map>
+#include <memory>
 #include "mesh/impl/RTreeAdapter.hpp"
 #include "mesh/Mesh.hpp"
 #include <boost/geometry.hpp>
-
+#include "mesh/Triangle.hpp"
+#include "mesh/Quad.hpp"
 
 // Forward declaration to friend the boost test struct
 namespace MeshTests {
 namespace RTree {
 struct CacheClearing;
 }}
 
-
 namespace precice {
 namespace mesh {
 
+/** The enumeration of various primitive types.
+ * \see PrimitiveIndex
+ * \see as_primitve_enum
+ */
+enum class Primitive {
+  Vertex,
+  Edge,
+  Triangle,
+  Quad
+};
+
+/// A standard print operator for Primitive
+std::ostream& operator<<(std::ostream& out, Primitive val);
+
+/// The type traits to return the enum value of a primitive type.
+template <class T>
+struct as_primitive_enum {
+};
+template <>
+struct as_primitive_enum<mesh::Vertex> {
+  static constexpr Primitive value = Primitive::Vertex;
+};
+template <>
+struct as_primitive_enum<mesh::Edge> {
+  static constexpr Primitive value = Primitive::Edge;
+};
+template <>
+struct as_primitive_enum<mesh::Triangle> {
+  static constexpr Primitive value = Primitive::Triangle;
+};
+template <>
+struct as_primitive_enum<mesh::Quad> {
+  static constexpr Primitive value = Primitive::Quad;
+};
+
+/** Binds an Index and Primitive into a type
+ * \see AABBGenerator
+ * \see indexPrimitives
+ */
+struct PrimitiveIndex {
+  Primitive type;
+  size_t    index;
+};
+
+/// Standard equality test for PrimitiveIndex
+bool operator==(const PrimitiveIndex& lhs, const PrimitiveIndex& rhs);
+
+/// Standard non-equality test for PrimitiveIndex
+bool operator!=(const PrimitiveIndex& lhs, const PrimitiveIndex& rhs);
+
+/// A standard print operator for PrimitiveIndex
+std::ostream& operator<<(std::ostream& out, PrimitiveIndex val);
+
+/// The axis aligned bounding box based on the Vertex Type
+using AABB = boost::geometry::model::box<Eigen::VectorXd>;
+
+/** Holding a reference to a Mesh it is a functor for packing
+ *  PrimitiveIndex into AABBs.
+ */
+class AABBGenerator
```

Comment:
Is there a chance to move that class to the implementation? Or even to `impl/` if it is purely internal use.

Reply:
I'll try to move some details things to `impl/RTree.hpp`